### PR TITLE
[mlir][EmitC] Add bitwise operators

### PR DESCRIPTION
**Pull request by marbre**
This adds operations for bitwise operators. Furthermore, an UnaryOp
class and a helper to print unary operations are introduced.

### CODE REVIEW

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td [hunk 1]
```diff
@@ -95,6 +95,118 @@ def EmitC_ApplyOp : EmitC_Op<"apply", []> {
...
 
+def EmitC_BitwiseAndOp : EmitC_BinaryOp<"bitwise_and", []> {
+  let summary = "Bitwise and operation";
+  let description = [{
+    With the `bitwise_and` operation the bitwise operator & (and) can
+    be applied.
+
+    Example:
+
+    ```mlirbool
```
Comment (simon-camp) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:106]:
```suggestion
    ```mlir
```

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 2]
```diff
@@ -588,6 +588,53 @@ static LogicalResult printOperation(CppEmitter &emitter,
...
+}
+
+static LogicalResult
+printOperation(CppEmitter &emitter,
+               emitc::BitwiseLeftShiftOp bitwiseLeftShiftOp) {
+  Operation *operation = bitwiseLeftShiftOp.getOperation();
+  return printBinaryOperation(emitter, operation, "<<");
+}
+
+static LogicalResult printOperation(CppEmitter &emitter,
```
Comment (simon-camp) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:604]: Can you add a `printUnaryOperation` helper and call it from here and the logical not op.

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td [hunk 3]
```diff
@@ -95,6 +95,118 @@ def EmitC_ApplyOp : EmitC_Op<"apply", []> {
...
+    %0 = emitc.bitwise_left_shift %arg0, %arg1 : (i32, i32) -> i32
+    ```
+    ```c++
+    // Code emitted for the operation above.
+    int32_t v3 = v1 << v2;
+    ```
+  }];
+}
+
+def EmitC_BitwiseNotOp : EmitC_Op<"bitwise_not", []> {
```
Comment (simon-camp) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:134]: Can you add an `EmitC_UnaryOp` for consistency and use it for the logical not as well.

**Review by simon-camp — CHANGES_REQUESTED** [review]
Looks good overall, when replacing binary with bitwise in the title and description.

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td (hunk 1 above)
Comment (marbre) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitC.td]: Good catch!

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td (hunk 3 above)
Comment (marbre) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitC.td]: Done :)

**Review by simon-camp — APPROVED** [review]

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp (hunk 2 above)
Comment (marbre) [reply mlir/lib/Target/Cpp/TranslateToCpp.cpp:620]: Yeah, already had this on the stack and had the intention to add is as follow up. Added it to this PR now and changed the description accordingly.

**Comment by marbre** [comment]
> Looks good overall, when replacing binary with bitwise in the title and description.

I obviously made too much use of the `EmitC_BinaryOp` class :laughing: Thanks for catching this one!

**Pull request merged**